src/concepts/lazy.md

```diff
@@ -0,0 +1,177 @@
+# Lazy Evaluation
+
+In programming languages, there's a few different strategies by which arguments
+of a function call are evaluated. One you should be familiar with is **_eager_**
+evaluation. SML is an **_eager_** language. In other words, we evaluate the
+arguments first. We'll be discussing how to "implement" the exact
+opposite form of evaluation: **_lazy_** evaluation. In **_lazy_** evaluation,
+arguments are evaluated as needed. To implement this, we'll make use of an idea
+called **thunks**.
+
+## Eager vs Lazy
+
+### Eager Evaluation
+
+Let's say I have the function `double : int -> int` which doubles a number, and
+`square : int -> int`, which squares a number. In both **_lazy_** and
+**_eager_** evaluation, `double square double 2 ==>* 32`. However, they differ
+in the way which they "arrive" at `32`.
+
+```sml
+val double : int -> int = fn n => 2 * n
+val squareE : int -> int = fn n => n * n
+
+(* Eager *)
+    double square double 2
+==> double square 4
+==> double 16
+==> 32
+```
+
+You can see that in an eager language, we evaluated the arguments first.
+However, in the lazy language, arguments will be evaluated as _needed_.
+The notion of _need_ will be subjective relative to what's being implemented.
+
+### Introducing Laziness with Thunks
+
+SML is an eager language. But what if we wanted to simulate lazy evaluation?
+First we need a way to wait until we _need_ an argument to evaluate it. This
+is where **thunks** come into play. A **thunk** is a function of type
+`unit -> 'a`. They help us represent this notion of "evaluation by need" by
```

Comment:
Should probably say that a thunk is a function of type `unit -> t` for some type `t`---this isn't the same as being a function of type `unit -> 'a` (of which there exist no total values).